### DISCUSSION TITLE
[Luigi Container] Share Pinia Store

### DISCUSSION THREAD

**Discussion by BossHogg97** [SAP/luigi#3588] [General]
Hi, i'm looking for a solution to share a Pinia store from Luigi Core app to Luigi Client. Both Core and Client app are developed with Vue3 framework.

Thanks

**Comment by ndricimrr** [comment]
Hi @BossHogg97 ,
Converted the issue to a discussion as it seems more suited for that instead.

I am not familiar with the Pinia store, however it seems like you are trying to transfer some data from Core to Client app. 
The type of framework/store you are using to manage your data/state management mechanism should not be a problem here (as long as it is Javascript based).

Therefore I could see this issue abstracted into simply Core -> Client communication. For this we have a dedicated section in our documentation that I encourage you to have a look at **[Core-Client Communication](https://docs.luigi-project.io/docs/communication)**

In essence, you are sending a message from the Core to the Client microfrontend (and back if necessary). 

Note that the particular **Core** -> **Client** communication mechanism in this case is based underneath on iframe-iframe message passing mechanism. (The client microfrontend/s in this case is hosted under an iframe wheres the Core here would be its parent) 

So whether you are using _Redux_ store in your React application, _ngrx/store_ for your Angular ones or _**Pinia store**_ for **VueJs**, in all cases this converges then into iframe message communication underneath. Looking at the examples in the documentation link above should give you an idea on how to approach your problem for sharing the state. 

Specific links and functions to get you started:
- [Luigi Core's **customMessages().sendToAll(...)**](https://docs.luigi-project.io/docs/luigi-core-api?section=custommessages)
- [Luigi Client's **addCustomMessageListener(...)**](https://docs.luigi-project.io/docs/luigi-client-api?section=addcustommessagelistener)
- [Luigi Client's **sendCustomMessage(...)**](https://docs.luigi-project.io/docs/luigi-client-api?section=sendcustommessage)

**Reply by BossHogg97** [reply]
Thanks for the answer. 

I'm not understanding where i have to write the **customMessages.sendToAll** function and how to call it from a specific vue component

**Reply by BossHogg97** [reply]
I'm using Luigi as container via package: "@luigi-project/container": "^1.0.0",

**Reply by ndricimrr** [reply]
The suggestion given above refers to if you are using **@luigi-project/core** together with **@luigi-project/client**.

If you are only using **@luigi-project/container** in your application it would be a different way of passing messages. 
You can check the respective [documentation for container here](https://docs.luigi-project.io/docs/luigi-container)

_More specific links to get you started from a given example:_
[App hosting Container sends message](https://github.com/SAP/luigi/blob/main/container/test-app/index2.html#L62)
[Container sends Message](https://github.com/SAP/luigi/blob/main/container/test-app/microfrontend.html#L74)

Would be nice to have some code snippets of what your environment looks like, since it makes it difficult to give suggestions this way. 

Hope this helps.

**Reply by BossHogg97** [reply]
In my environment i have the following component that i use for visualize the micro-frontend.
```
<script setup lang="ts">
  import '@luigi-project/container'
  import { MfeEnum } from '@/enums'
  import { keycloak } from '@xxx-apps/shared/src/composables/useKeycloak'

  // Store Import
  import { useMfeStore } from '@xxx-apps/shared/src/stores/app/mfeStore'
  import { storeToRefs } from 'pinia'

  interface Props {
    url: string
    mfeName: MfeEnum
  }
  const props = defineProps<Props>()

  // Define context obj
  let contextObj =  {
        plantId: plantId.value,
        token: token.value,
        productionId: productionId.value,
        isErmOpenFromProduction: isErmOpenFromProduction.value,
  }
  const contextJson = JSON.stringify(contextObj)
</script>

<template>
  <div class="mfe-container">
    <n-card :content-style="{ padding: 0 }" class="h-full" id="col1">
      <luigi-container :viewURL="props.url" :context="contextJson" />
    </n-card>
  </div>
</template>

<style scoped>
  .mfe-container {
    height: calc(100vh - 120px);
    /* border: 1px solid rgb(38, 0, 255); */
  }
</style>
```
The above code is executed when i click on btn (inside host app) and allow me to display the MFE. I also set a context object and all work fine. In my situation i have keycloak and i need to update the token when i refresh it inside host application. The token is actually set in context object only once at startup of luigi-container.

For these reason i need a way to update this information when the token is refreshed.

I hope this explanation is usefull for understand the context.

**Reply by ndricimrr** [reply]
I see. 
I think in this case context is first set initially on startup of luigi-container. 

In your situation it would make sense to just update the context post-startup of luigi-container instead of sending custom messages.

First get a handle of that particular luigi-container element:
If I am not mistaken you can take a reference of the element in VueJS like so:
```
<luigi-container ref="luigiContainer" :viewURL="props.url" :context="contextJson" />
 const myLuigiContainer = this.$refs.luigiContainer;
```

I believe you must have a place where you check if token value has changed/refreshed: lets call that: onTokenValueChanged()

In your case you could do the following by acting on token value refreshing:
```
onTokenValueChanged(newToken) {
   myLuigiContainer.updateContext({
     // pass the new object with the newToken here 
   })
}
```

**Reply by clabnet** [reply]
I'm very interested to this case. I would use Pinia with Luigi.

**Reply by BossHogg97** [reply]
Great, i'll try as soon as possible and let you know!

Thanks for the amazing support!

**Reply by BossHogg97** [reply]
I tried your solution and it work fine. In the Luigi Client i used a **setInterval** for check possible context update. Is there a better way or Luigi Client API that look for context update?
```
setInterval(() => {
  const receivedContext1 = LuigiClient.getContext()
  console.log('receivedContext1: ', receivedContext1)
}, 5001)
```

On Core application i update the context only once i receive an event with new token

```
// Define variable for get component ref
const myLuigiContainer = ref(null)

evtUpdateContext.on((newToken) => {
    myLuigiContainer.value.updateContext({
      token: newToken,
    })
  })
```

**Reply by ndricimrr** [reply]
Hi, 
You can use a [context update listener](https://docs.luigi-project.io/docs/luigi-client-api/?section=addcontextupdatelistener) on the Luigi Client side:
// you can use the id in case you ever need to [remove the listener](https://docs.luigi-project.io/docs/luigi-client-api/?section=removecontextupdatelistener)
```
const updateListenerId = LuigiClient.addContextUpdateListener((context) => {
        console.log('receivedContext: ', context)
        // do something with received context
})
```

**Reply by BossHogg97** [reply]
It work! Thanks!!

Just for know. Is there a place where there is a list of available function usable with container reference?

**Reply by ndricimrr** [reply]
Glad that helped. We're in the process of polishing that list up into the documentation, but at the moment you may have a look [here](https://github.com/SAP/luigi/blob/main/container/typings/LuigiContainer.svelte.d.ts)

**Reply by BossHogg97** [reply]
Very nice! I saw that the doc related to container "version" is small at the moment :(
